### PR TITLE
Rules

### PR DESCRIPTION
**Pull request by EliasEriksson**
# Rules
Rules comes in as a string on format `type time selectivelyString` which can be parsed and used to determine the expected time to work on a date (see tests in `/Rule/index.spec.ts`). 

## Types of rules
Rules comes in three different types: 
* Set with absolute time
* Adjust with absolute time
* Adjust with percentage

## Calculating expected time
Rules are processed from an array in the order low to high index. Each rule modifies the time from the previous rule (initial value is nothing) and the expected time is the end result.

### CODE REVIEW

File: Rule/State.spec.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import { isoly } from "isoly"
+import { State } from "./State"
+
+describe("Rule.State", () => {
+	it("is", () => {
+		const state: State = {
+			date: "2023-01-01",
+			year: "2023",
```
Comment (LinusDahlstrom) [inline Rule/State.spec.ts:8]: I think this has been up for discussion several times and this comment is just to refresh my memory. We did conclude that we require to double save the year in both the year property and the date property. Right?

File: Rule/State.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	})
+	export const is = type.is
+	export const flaw = type.flaw
+	export function create(date: isoly.Date, user: userwidgets.Email): State {
+		return {
+			date,
+			year: isoly.Date.getYear(date).toString(10),
+			month: isoly.Date.getMonth(date).toString(10),
+			week: isoly.Date.getWeek(date).toString(10),
+			day: isoly.Date.getDay(date).toString(10),
```
Comment (LinusDahlstrom) [inline Rule/State.ts:32]: what does the toString(10) accomplish?

File: Rule/Time/Percentage.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+			const percentage = parseFloat(percentageString)
+			if (!Base.safe(percentage))
+				result = undefined
+			else
+				result = percentage / 100
+		}
+		return result
+	}
+	export function toString(time: Percentage): string {
+		return `${(time * 100).toString(10)}%`
```
Comment (LinusDahlstrom) [inline Rule/Time/Percentage.ts:24]: I left a comment elsewhere asking about toString(10) but I found the documentation now and it seems like it is to tell which base to use when writing out the number. Base 10 appears to be the default value so then we wouldnt need to write out 10.

File: Rule/Time/Span.spec.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+		expect(Span.parse("-8:50h")).toEqual({ hours: -8, minutes: -50 })
+		expect(Span.parse("-8:h")).toEqual({ hours: -8, minutes: undefined })
+		expect(Span.parse("-8:00h")).toEqual({ hours: -8, minutes: undefined })
+		expect(Span.parse("-888:00h")).toEqual({ hours: -888, minutes: undefined })
+		expect(Span.parse("-0:30h")).toEqual({ hours: -0, minutes: -30 })
+		expect(Span.parse("-0:3h")).toEqual({ hours: -0, minutes: -3 })
+		expect(Span.parse("-1:45h")).toEqual({ hours: -1, minutes: -45 })
+		expect(Span.parse("-666:666h")).toEqual({ hours: -666, minutes: -666 })
+		// failing
+		expect(Span.parse("8:50")).toEqual(undefined)
```
Comment (LinusDahlstrom) [inline Rule/Time/Span.spec.ts:30]: why is the result of this test undefined?

**Review by LinusDahlstrom — CHANGES_REQUESTED** [review]

File: Rule/State.ts (hunk 2 above)
Comment (LinusDahlstrom) [reply Rule/State.ts:32]: it converts the number to a string using base 10.
Base 10 is the default behaivour, do we still want to type it out explicitly?

File: Rule/State.spec.ts (hunk 1 above)
Comment (EliasEriksson) [reply Rule/State.spec.ts:8]: The state is the state on which selectively operate. The reason why both date and year is provided is to provide more options for selectively rules to target more easely.

File: Rule/State.ts (hunk 2 above)
Comment (EliasEriksson) [reply Rule/State.ts:32]: We do not, but i like to be explicit sometimes :upside_down_face:

File: Rule/Time/Percentage.ts (hunk 3 above)
Comment (EliasEriksson) [reply Rule/Time/Percentage.ts:24]: It is indeed. I have just preferred to be explicit.

File: Rule/Time/Span.spec.ts (hunk 4 above)
Comment (EliasEriksson) [reply Rule/Time/Span.spec.ts:30]: There is no unit. Either `h` or `%` is required to be able to parse a time. In this case its a `Time.Span` that is parsing and is expecting a trailing `h` for the parse to succeed.

**Review by LinusDahlstrom — APPROVED** [review]

**Pull request merged**